### PR TITLE
dlpack support

### PR DESCRIPTION
**Pull request by VoVAllen**

### CODE REVIEW

File: tensorflow/c/eager/dlpack.cc [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+const Tensor* GetTensorFromHandle(TFE_TensorHandle* h, TF_Status* status) {
+  TensorHandle* handle = GetTensorHandleFromTFEHandle(h, status);
+
+  if (handle->IsRemote()) {
+    status->status = tensorflow::errors::InvalidArgument(
+        "TFE_TensorHandleDevicePointer may not be called on a remote tensor "
+        "handle.");
+    return nullptr;
+  }
```
Comment (jermainewang) [inline tensorflow/c/eager/dlpack.cc]: redundant?

File: tensorflow/c/eager/dlpack.cc [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+  tensorflow::Device* device(absl::get<tensorflow::Device*>(handle->device()));
+  if (device != nullptr) {
+    status->status = device->Sync();
```
Comment (jermainewang) [inline tensorflow/c/eager/dlpack.cc]: Should device be synced here? In RFC discussion, synchronization is performed when a TF tensor is converted to a DLPack tensor.

File: tensorflow/c/eager/dlpack.cc [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+  return tensor;
+};
+
+void deleter(DLManagedTensor* arg) {
+  TFDLMTensor* owner = static_cast<TFDLMTensor*>(arg->manager_ctx);
+  owner->handle->Unref();
+  delete owner;
+}
+
+DLDataType getDLDataType(TF_DataType data_type, TF_Status* status) {
```
Comment (jermainewang) [inline tensorflow/c/eager/dlpack.cc]: `GetDLDataType`

File: tensorflow/c/eager/dlpack.cc [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+  }
+  const tensorflow::Tensor* tensor;
+  status->status = handle->Tensor(&tensor);
+  if (!status->status.ok()) {
+    return nullptr;
+  }
+  return tensor;
+};
+
+void deleter(DLManagedTensor* arg) {
```
Comment (jermainewang) [inline tensorflow/c/eager/dlpack.cc]: `DLManagedTensorDeleter`

File: tensorflow/c/eager/dlpack.cc [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          "Unsupported TF_DataType is not supported by dlpack");
+      break;
+  }
+  return dtype;
+}
+
+DLContext getDLContext(TFE_TensorHandle* h, TF_Status* status) {
```
Comment (jermainewang) [inline tensorflow/c/eager/dlpack.cc]: `GetDLContext`

File: tensorflow/c/eager/dlpack.cc [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+  DLContext ctx;
+  const char* device_name = h->handle->DeviceName(&status->status);
+  DeviceNameUtils::ParsedName parsed_name;
+  tensorflow::DeviceNameUtils::ParseFullName(absl::string_view(device_name),
+                                             &parsed_name);
+  std::string device_type = parsed_name.type;
+  int device_id = -1;
+  if (parsed_name.has_id) {
+    device_id = parsed_name.id;
+  }  // Question? device_id?=-1
```
Comment (jermainewang) [inline tensorflow/c/eager/dlpack.cc]: throw error?

File: tensorflow/python/tfe_wrapper.cc [hunk 7]
```diff
@@ -1033,6 +1034,50 @@ PYBIND11_MODULE(_pywrap_tfe, m) {
...
+        tensorflow::make_safe(TF_NewStatus());
+    if (absl::string_view(pycapsule.name()) !=
+        tensorflow::kDlTensorCapsuleName) {
+      status->status = tensorflow::errors::InvalidArgument(
+          "DLPack tensor must be a capsule with name \"dltensor\", got \"%s\". "
+          "Note that a DLPack tensor may be consumed at most once.",
+          absl::string_view(pycapsule.name()));
+    }
+    TFE_TensorHandle* thandle = tensorflow::TFE_HandleFromDLPack(
+        static_cast<void*>(pycapsule), status.get());
```
Comment (jermainewang) [inline tensorflow/python/tfe_wrapper.cc]: No need of `static_cast` for `void*`

**Review by jermainewang — COMMENTED** [review]
Overall the code is good to me. I found you've already pushed a PR to TF repo. Let's wait for their response.  Several questions:
* According to the RFC, synchronization will be performed after ToDLPack. Is this correctly implemented?
* I didn't see any python codes. Has this been tested?

File: tensorflow/c/eager/dlpack.cc (hunk 2 above)
Comment (VoVAllen) [reply tensorflow/c/eager/dlpack.cc]: You are right. It's copyed from another function and I didn't carefully check it.

**Pull request closed without merging**